### PR TITLE
Bug 1829385: fix alignment issues with MultiColumnField in Traffic Splitting modal

### PR DESCRIPTION
**Pull request by vikram-raj**
**Fixes**: 
https://issues.redhat.com/browse/ODC-3421

**Analysis / Root cause**: 
MultiColumnField in Traffic Splitting modal is not aligned properly.

**Solution Description**: 
Align the MultiColumnField in Traffic Splitting modal by fixing the width of the field.

**Screen shots / Gifs for design review**: 
traffic

**Unit test coverage report**: 
Screenshot 2020-05-06 at 11 22 40 AM

**Browser conformance**: 

- [x] Chrome
- [x] Firefox
- [x] Safari
- [ ] Edge

### CODE REVIEW

**Comment by vikram-raj** [comment]
/retitle Bug 1829385: fix alignment issues with MultiColumnField in Traffic Splitting modal

**Comment by vikram-raj** [comment]
cc: @openshift/team-devconsole-ux

File: frontend/packages/knative-plugin/src/components/traffic-splitting/TrafficSplittingModal.scss [hunk 1]
```diff
@@ -0,0 +1,6 @@
+.odc-traffic-splitting-modal {
+  .odc-multi-column-field__col {
+    flex: auto;
+    width: 30%;
```
Comment (andrewballantyne) [inline frontend/packages/knative-plugin/src/components/traffic-splitting/TrafficSplittingModal.scss:4]:
```suggestion
    flex-grow: 1;
```
Grow to the same size? I haven't tested this out, not sure if the `grow-basis` will impact the solution, may need `grow: 1 1 0;`

Comment (vikram-raj) [reply frontend/packages/knative-plugin/src/components/traffic-splitting/TrafficSplittingModal.scss:4]: I tried ```flex-grow``` and it does work in this case as the width of the dropdown varies as per the option length. That is why it is not aligned.

Comment (andrewballantyne) [reply frontend/packages/knative-plugin/src/components/traffic-splitting/TrafficSplittingModal.scss:4]: Flex-box in the horizontal direction handles width. You don't need the width property. If it's not working you're conflicting with other styles / indeterminate items. I'll look more into it in a few, haven't checked out the code yet.

File: frontend/packages/knative-plugin/src/components/traffic-splitting/TrafficSplittingModal.scss [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    flex: none;
+    width: 10%;
+  }
+  .odc-multi-column-field__col:nth-of-type(2) {
+    width: 20%;
+  }
+  .odc-multi-column-field__col:nth-of-type(3) {
+    width: 60%;
+  }
+}
```
Comment (andrewballantyne) [inline frontend/packages/knative-plugin/src/components/traffic-splitting/TrafficSplittingModal.scss:12]: This is not an appropriate solution @vikram-raj - this is a hack at best. For one, your percentages don't equal 100. 

I think we need to update the `MultiColumnField` component to support this logic between "splitting the space equally with everyone". Otherwise we'll end up with this problem on every use of this component moving forward as well.

**Comment by christianvogt** [comment]
@vikram-raj seems to me that this can occur with any usage of the `MultiColumnField` and therefore would be better off having a solution that can be reused everywhere.
I think what we want is the ability to use the standard 12 column distribution.
Would be good to even support a responsive layout in future.

Why don't we allow the for the children of the component to specify the grid spans?
eg.
```
<MultiColumnFieldCell span={2}><InputField /></MultiColumnFieldCell>
```

By default we evenly distribute the spans based on the number of columns.

cc @rohitkrai03  thoughts?

**Comment by rohitkrai03** [comment]
> @vikram-raj seems to me that this can occur with any usage of the `MultiColumnField` and therefore would be better off having a solution that can be reused everywhere.
> I think what we want is the ability to use the standard 12 column distribution.
> Would be good to even support a responsive layout in future.
> 
> Why don't we allow the for the children of the component to specify the grid spans?
> eg.
> 
> ```
> <MultiColumnFieldCell span={2}><InputField /></MultiColumnFieldCell>
> ```
> 
> By default we evenly distribute the spans based on the number of columns.
> 
> cc @rohitkrai03 thoughts?

I agree with @christianvogt here, we should look at finding the solution which fixes this issue with all the usages of this component. I like the idea of children defining the space they need based on the 12 column grid layout. By default all columns can be divided into equal space.

**Comment by vikram-raj** [comment]
Added pf `Grid` layout to ```MultiColumnField``` component.

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss [hunk 3]
```diff
@@ -1,15 +1,12 @@
 .odc-multi-column-field {
   &__row {
-    display: flex;
-    flex-direction: row;
     position: relative;
     margin-right: var(--pf-global--spacer--lg);
     margin-bottom: var(--pf-global--spacer--lg);
     max-height: var(--pf-global--spacer--xl);
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss:6]: Just saw that we have a max height per row. unsure why that is. @rohitkrai03 ?

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss [hunk 4]
```diff
@@ -18,6 +15,7 @@
     line-height: var(--pf-global--spacer--xl);
     position: absolute;
     right: -30px;
+    top: 0;
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss:18]: You'll want want to center the button vertically. For a typical row, the result is the same, but gives more flexibility if we start creating rows where the height differs from the button height.
```
top: 50%;
transform: translateY(-50%);
```

This would enable us to support the same layout as is done here:
image

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnFieldHeader.tsx [hunk 5]
```diff
@@ -1,18 +1,22 @@
...
+const MultiColumnFieldHeader: React.FC<MultiColumnFieldHeaderProps> = ({ headers, span = [] }) => (
   <div className="odc-multi-column-field__row">
-    {headers.map((header) => (
-      <div className="odc-multi-column-field__col" key={header}>
-        {header}
-      </div>
-    ))}
-    <div className="odc-multi-column-field__col--button" />
+    <Grid className="odc-multi-column-field__row">
+      {headers.map((header, key) => (
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnFieldHeader.tsx:13]: `key` => `i` or `index`

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnFieldRow.tsx [hunk 6]
```diff
@@ -51,17 +60,22 @@ const MultiColumnFieldRow: React.FC<MultiColumnFieldRowProps> = ({
...
-      const fieldName = `${name}.${rowIndex}.${child.props.name}`;
-      const newProps = { ...child.props, name: fieldName };
-      return (
-        <div key={fieldName} className="odc-multi-column-field__col">
-          {React.cloneElement(child, newProps)}
-        </div>
-      );
-    })}
+    <Grid>
+      {React.Children.map(children, (child: React.ReactElement, key) => {
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnFieldRow.tsx:67]: `key` => `i` or `index`

File: frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/PipelineParameters.tsx [hunk 7]
```diff
@@ -20,6 +20,7 @@ const PipelineParameters: React.FC<PipelineParametersProps> = (props) => {
       emptyValues={{ name: '', description: '', default: '' }}
       emptyMessage={emptyMessage}
       isReadOnly={isReadOnly}
+      span={[4, 4, 4]}
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/PipelineParameters.tsx:23]: Shouldn't be required and therefore columns should be evenly spaced.
Same comment for all instances where columns should be evenly spaced.

File: frontend/packages/knative-plugin/src/components/traffic-splitting/TrafficSplittingModal.tsx [hunk 8]
```diff
@@ -23,7 +23,7 @@ const TrafficSplittingModal: React.FC<Props> = ({
   values,
 }) => {
   return (
-    <form className="modal-content" onSubmit={handleSubmit}>
+    <form className="modal-content odc-traffic-splitting-modal" onSubmit={handleSubmit}>
```
Comment (christianvogt) [inline frontend/packages/knative-plugin/src/components/traffic-splitting/TrafficSplittingModal.tsx:26]: I don't see `odc-traffic-splitting-modal` defined.

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss (hunk 4 above)
Comment (andrewballantyne) [reply frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss:18]: Would not just a flex wrapper fix this? There are align center options natively available without going to negative translations.

File: frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/PipelineParameters.tsx (hunk 7 above)
Comment (christianvogt) [reply frontend/packages/dev-console/src/components/pipelines/detail-page-tabs/PipelineParameters.tsx:23]: Not that we will be creating such layout but simply make them all even then add 1 to each column starting from the beginning until you've used up all the spaces.
eg.
5 = [3,3,2,2,2]
7 = [2,2,2,2,2,1,1]
9 = [2,2,2,1,1,1,1,1,1]

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss (hunk 3 above)
Comment (andrewballantyne) [reply frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss:6]: Ping @rohitkrai03 ^

Comment (rohitkrai03) [reply frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss:6]: I don't exactly remember as to why I added this but I think it was something to do with the UX designs at the moment and rows being bigger than the designs.

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss [hunk 9]
```diff
@@ -1,23 +1,24 @@
...
   &__col--button {
-    cursor: pointer;
     line-height: var(--pf-global--spacer--xl);
     position: absolute;
     right: -30px;
+    top: 50%;
+    transform: translateY(-50%);
+    button:not(:disabled) {
+      cursor: pointer;
+    }
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss:21]: Why is this needed when we are using PF Button which already has a pointer cursor?

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/multicolumn-field-utils.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  while (_.sum(spans) !== 12) {
+    for (let i = 0; i < totalFieldCount; i++) {
+      spans[i]++;
+      if (_.sum(spans) === 12) {
+        break;
+      }
+    }
+  }
+  return spans;
+};
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/formik-fields/multi-column-field/multicolumn-field-utils.ts:21]: This can be simplified to something like this:
```
const spans = _.fill(Array(totalFieldCount), Math.trunc(12 / totalFieldCount));
let remainder = 12 % totalFieldCount;
while (remainder--) {
  spans[remainder]++;
}
return spans;
```

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/multicolumn-field-utils.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import * as _ from 'lodash';
+import { gridItemSpanValueShape } from '@patternfly/react-core';
+
+export const getSpans = (totalFieldCount: number): gridItemSpanValueShape[] => {
```
Comment (christianvogt) [inline frontend/packages/console-shared/src/components/formik-fields/multi-column-field/multicolumn-field-utils.ts:4]:
```suggestion
export const getSpans = (totalFieldCount: gridItemSpanValueShape): gridItemSpanValueShape[] => {
```

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss (hunk 9 above)
Comment (andrewballantyne) [reply frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.scss:21]: 🤔 good point... I just moved the cursor without testing if it was needed.

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.tsx [hunk 12]
```diff
@@ -23,9 +24,14 @@ const MultiColumnField: React.FC<MultiColumnFieldProps> = ({
   disableDeleteRow,
   disableAddRow,
   toolTip,
+  spans,
 }) => {
   const { values } = useFormikContext<FormikValues>();
   const fieldValue = _.get(values, name, []);
+  const totalFieldCount: gridItemSpanValueShape = React.Children.count(
+    children,
+  ) as gridItemSpanValueShape;
```
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.tsx:33]: Typing and casting should not be needed... did TypeScript yell at you?

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/multicolumn-field-utils.ts [hunk 13]
```diff
@@ -0,0 +1,17 @@
+import * as _ from 'lodash';
+import { gridItemSpanValueShape } from '@patternfly/react-core';
+
+export const getSpans = (totalFieldCount: gridItemSpanValueShape): gridItemSpanValueShape[] => {
+  const spans: gridItemSpanValueShape[] = _.fill(
+    Array(totalFieldCount),
+    Math.trunc(12 / totalFieldCount),
+  ) as gridItemSpanValueShape[];
```
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/components/formik-fields/multi-column-field/multicolumn-field-utils.ts:8]: Same here, is the type & cast needed by TypeScript?

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.tsx (hunk 12 above)
Comment (vikram-raj) [reply frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.tsx:33]: yes, after changing the type of ```totalFieldCount``` to ```gridItemSpanValueShape``` from ```number``` in the utils.

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/multicolumn-field-utils.ts (hunk 13 above)
Comment (vikram-raj) [reply frontend/packages/console-shared/src/components/formik-fields/multi-column-field/multicolumn-field-utils.ts:8]: Yes, I was getting the type error because after adding return type ```gridItemSpanValueShape[]``` to this utils.

Comment (andrewballantyne) [reply frontend/packages/console-shared/src/components/formik-fields/multi-column-field/multicolumn-field-utils.ts:8]: Ah right... because `_.fill` has probably a very generic type (if any). Well, it's a nit at this point, but you don't have to type the variable if you're casting. It's always _that_ type.

File: frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.tsx (hunk 12 above)
Comment (christianvogt) [reply frontend/packages/console-shared/src/components/formik-fields/multi-column-field/MultiColumnField.tsx:33]: Sorry, i caused this because I didn't follow the code while making the suggestion for the input param type.
Don't care that much which approach you take so long as the `getSpans` function supports a number > 12.
